Connection issues to database after long running period

**Describe the bug**
After a long-running period of the server, the plugin cant get or update any SQL data. Therefore its not possible to join the server.

**To Reproduce**
Steps to reproduce the behavior:
1. Let the server running for an extended period.
2. Try to join the server.

**Expected behavior**
Even after a long period I should be able to join the server.

**Additional context**
```
[11:12:39 ERROR] [org.kryonite.kryodiscordwhitelist.velocity.listener.PlayerListener]: Failed to get or update player
java.sql.SQLNonTransientConnectionException: (conn=9583) Connection reset by peer
        at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.createException(ExceptionFactory.java:73) ~[?:?]
        at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.create(ExceptionFactory.java:158) ~[?:?]
        at org.mariadb.jdbc.MariaDbStatement.executeExceptionEpilogue(MariaDbStatement.java:266) ~[?:?]
        at org.mariadb.jdbc.ClientSidePreparedStatement.executeInternal(ClientSidePreparedStatement.java:229) ~[?:?]
        at org.mariadb.jdbc.ClientSidePreparedStatement.execute(ClientSidePreparedStatement.java:149) ~[?:?]
        at org.mariadb.jdbc.ClientSidePreparedStatement.executeQuery(ClientSidePreparedStatement.java:163) ~[?:?]
        at org.kryonite.kryodiscordwhitelist.common.persistence.repository.impl.MariaDbUserRepository.get(MariaDbUserRepository.java:101) ~[?:?]
        at org.kryonite.kryodiscordwhitelist.velocity.listener.PlayerListener.onPlayerLogin(PlayerListener.java:24) ~[?:?]
        at org.kryonite.kryodiscordwhitelist.velocity.listener.Lmbda$10.execute(Unknown Source) ~[?:?]
        at com.velocitypowered.proxy.event.UntargetedEventHandler$VoidHandler.lambda$buildHandler$0(UntargetedEventHandler.java:47) ~[velocity.jar:3.1.2-SNAPSHOT (git-b8f1df44-b110)]
        at com.velocitypowered.proxy.event.VelocityEventManager.fire(VelocityEventManager.java:598) ~[velocity.jar:3.1.2-SNAPSHOT (git-b8f1df44-b110)]
        at com.velocitypowered.proxy.event.VelocityEventManager.lambda$fire$5(VelocityEventManager.java:479) ~[velocity.jar:3.1.2-SNAPSHOT (git-b8f1df44-b110)]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: org.mariadb.jdbc.internal.util.exceptions.MariaDbSqlException: Connection reset by peer
        at org.mariadb.jdbc.internal.util.exceptions.MariaDbSqlException.of(MariaDbSqlException.java:34) ~[?:?]
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.exceptionWithQuery(AbstractQueryProtocol.java:194) ~[?:?]
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.exceptionWithQuery(AbstractQueryProtocol.java:177) ~[?:?]
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:323) ~[?:?]
        at org.mariadb.jdbc.ClientSidePreparedStatement.executeInternal(ClientSidePreparedStatement.java:220) ~[?:?]
        ... 11 more
Caused by: java.sql.SQLNonTransientConnectionException: Connection reset by peer
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.handleIoException(AbstractQueryProtocol.java:2090) ~[?:?]
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:323) ~[?:?]
        at org.mariadb.jdbc.ClientSidePreparedStatement.executeInternal(ClientSidePreparedStatement.java:220) ~[?:?]
        ... 11 more
Caused by: java.net.SocketException: Connection reset by peer
        at sun.nio.ch.NioSocketImpl.implWrite(NioSocketImpl.java:420) ~[?:?]
        at sun.nio.ch.NioSocketImpl.write(NioSocketImpl.java:440) ~[?:?]
        at sun.nio.ch.NioSocketImpl$2.write(NioSocketImpl.java:826) ~[?:?]
        at java.net.Socket$SocketOutputStream.write(Socket.java:1035) ~[?:?]
        at org.mariadb.jdbc.internal.io.output.StandardPacketOutputStream.flushBuffer(StandardPacketOutputStream.java:111) ~[?:?]
        at org.mariadb.jdbc.internal.io.output.AbstractPacketOutputStream.flush(AbstractPacketOutputStream.java:186) ~[?:?]
        at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:316) ~[?:?]
        at org.mariadb.jdbc.ClientSidePreparedStatement.executeInternal(ClientSidePreparedStatement.java:220) ~[?:?]
        ... 11 more
```

--- a/bot/src/main/java/org/kryonite/kryodiscordwhitelist/bot/KryoDiscordWhitelistBot.java
+++ b/bot/src/main/java/org/kryonite/kryodiscordwhitelist/bot/KryoDiscordWhitelistBot.java
@@ -1,7 +1,7 @@
 package org.kryonite.kryodiscordwhitelist.bot;
 
-import java.sql.Connection;
-import java.sql.DriverManager;
+import com.zaxxer.hikari.HikariConfig;
+import com.zaxxer.hikari.HikariDataSource;
 import java.sql.SQLException;
 import javax.security.auth.login.LoginException;
 import net.dv8tion.jda.api.JDABuilder;
@@ -18,13 +18,15 @@ public class KryoDiscordWhitelistBot extends ListenerAdapter {
   public KryoDiscordWhitelistBot() throws SQLException {
     setupForkJoinPoolParallelism();
 
-    Connection connection = DriverManager.getConnection(getEnv("CONNECTION_STRING"));
-    userRepository = new MariaDbUserRepository(connection);
+    HikariConfig hikariConfig = new HikariConfig();
+    hikariConfig.setJdbcUrl(getEnv("CONNECTION_STRING"));
+
+    userRepository = new MariaDbUserRepository(new HikariDataSource(hikariConfig));
   }
 
-  public KryoDiscordWhitelistBot(Connection connection) throws SQLException {
+  public KryoDiscordWhitelistBot(HikariDataSource hikariDataSource) throws SQLException {
     setupForkJoinPoolParallelism();
-    userRepository = new MariaDbUserRepository(connection);
+    userRepository = new MariaDbUserRepository(hikariDataSource);
   }
 
   public static void main(String[] args) throws LoginException, InterruptedException, SQLException {

--- a/bot/src/main/java/org/kryonite/kryodiscordwhitelist/bot/listener/MessageListener.java
+++ b/bot/src/main/java/org/kryonite/kryodiscordwhitelist/bot/listener/MessageListener.java
@@ -1,6 +1,7 @@
 package org.kryonite.kryodiscordwhitelist.bot.listener;
 
 import java.sql.SQLException;
+import java.sql.SQLIntegrityConstraintViolationException;
 import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
 import net.dv8tion.jda.api.entities.Message;
@@ -27,38 +28,50 @@ public class MessageListener extends ListenerAdapter {
     Message message = event.getMessage();
     String messageContent = message.getContentStripped();
 
-    boolean saveSuccessful = saveUserToDatabase(event, messageContent);
+    SaveUserResult saveUserResult = saveUserToDatabase(event, messageContent);
 
     message.delete()
-        .and(event.getChannel().sendMessage(createReply(event, messageContent, saveSuccessful)))
+        .and(event.getChannel().sendMessage(createReply(event, messageContent, saveUserResult)))
         .queue();
   }
 
-  private boolean saveUserToDatabase(@NotNull MessageReceivedEvent event, String messageContent) {
+  private SaveUserResult saveUserToDatabase(@NotNull MessageReceivedEvent event, String messageContent) {
     User user = User.create(event.getAuthor().getIdLong(), messageContent, null);
     log.info("{}", user);
     try {
       userRepository.save(user);
     } catch (SQLException exception) {
+      if (exception instanceof SQLIntegrityConstraintViolationException) {
+        return SaveUserResult.DUPLICATE;
+      }
+
       log.error("Failed to save user {}", user, exception);
-      return false;
+      return SaveUserResult.FAILED;
     }
 
-    return true;
+    return SaveUserResult.SUCCESSFUL;
   }
 
   private String createReply(@NotNull MessageReceivedEvent event, String messageContent,
-                             boolean saveSuccessful) {
+                             SaveUserResult saveUserResult) {
     StringBuilder reply = new StringBuilder()
         .append("<@")
         .append(event.getAuthor().getIdLong())
         .append("> ")
         .append(messageContent);
-    if (saveSuccessful) {
-      reply.append(" wurde erfolgreich auf die Whitelist gesetzt.");
-    } else {
-      reply.append(" konnte nicht erfolgreich auf die Whitelist gesetzt werden.");
+
+    switch (saveUserResult) {
+      case SUCCESSFUL -> reply.append(" wurde erfolgreich auf die Whitelist gesetzt.");
+      case DUPLICATE -> reply.append(" befindet sich bereits auf der Whitelist.");
+      case FAILED -> reply.append(" konnte nicht erfolgreich auf die Whitelist gesetzt werden.");
+      default -> throw new IllegalStateException("Unexpected value: " + saveUserResult);
     }
     return reply.toString();
+  }
+
+  private enum SaveUserResult {
+    SUCCESSFUL,
+    DUPLICATE,
+    FAILED
   }
 }

--- a/bot/src/test/java/org/kryonite/kryodiscordwhitelist/bot/KryoDiscordWhitelistBotTest.java
+++ b/bot/src/test/java/org/kryonite/kryodiscordwhitelist/bot/KryoDiscordWhitelistBotTest.java
@@ -2,7 +2,7 @@ package org.kryonite.kryodiscordwhitelist.bot;
 
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 
-import java.sql.Connection;
+import com.zaxxer.hikari.HikariDataSource;
 import java.sql.SQLException;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
@@ -14,12 +14,12 @@ import org.mockito.junit.jupiter.MockitoExtension;
 class KryoDiscordWhitelistBotTest {
 
   @Mock(answer = Answers.RETURNS_DEEP_STUBS)
-  private Connection connection;
+  private HikariDataSource hikariDataSource;
 
   @Test
   void shouldSetupInstance() throws SQLException {
     // Arrange - Act
-    KryoDiscordWhitelistBot result = new KryoDiscordWhitelistBot(connection);
+    KryoDiscordWhitelistBot result = new KryoDiscordWhitelistBot(hikariDataSource);
 
     // Assert
     assertNotNull(result, "Result should not be null");

--- a/bot/src/test/java/org/kryonite/kryodiscordwhitelist/bot/listener/MessageListenerTest.java
+++ b/bot/src/test/java/org/kryonite/kryodiscordwhitelist/bot/listener/MessageListenerTest.java
@@ -1,6 +1,7 @@
 package org.kryonite.kryodiscordwhitelist.bot.listener;
 
 import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.argThat;
 import static org.mockito.Mockito.doThrow;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.never;
@@ -8,6 +9,8 @@ import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 import java.sql.SQLException;
+import java.sql.SQLIntegrityConstraintViolationException;
+import net.dv8tion.jda.api.entities.Message;
 import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
@@ -38,6 +41,26 @@ class MessageListenerTest {
     when(messageReceivedEvent.getAuthor().getIdLong()).thenReturn(discordId);
     when(messageReceivedEvent.getChannel().getName()).thenReturn("whitelist");
     when(messageReceivedEvent.getMessage().getContentStripped()).thenReturn(discordMessage);
+
+    // Act
+    testee.onMessageReceived(messageReceivedEvent);
+
+    // Assert
+    verify(userRepositoryMock).save(User.create(discordId, discordMessage, null));
+  }
+
+  @Test
+  void shouldNotPersistUser_WhenAlreadyOnWhitelist() throws SQLException {
+    // Arrange
+    long discordId = 123456789123L;
+    String discordMessage = "Test1234";
+
+    MessageReceivedEvent messageReceivedEvent = mock(MessageReceivedEvent.class, Answers.RETURNS_DEEP_STUBS);
+    when(messageReceivedEvent.getAuthor().isBot()).thenReturn(false);
+    when(messageReceivedEvent.getAuthor().getIdLong()).thenReturn(discordId);
+    when(messageReceivedEvent.getChannel().getName()).thenReturn("whitelist");
+    when(messageReceivedEvent.getMessage().getContentStripped()).thenReturn(discordMessage);
+    doThrow(SQLIntegrityConstraintViolationException.class).when(userRepositoryMock).save(any());
 
     // Act
     testee.onMessageReceived(messageReceivedEvent);

--- a/common/build.gradle.kts
+++ b/common/build.gradle.kts
@@ -1,3 +1,4 @@
 dependencies {
     api("org.mariadb.jdbc:mariadb-java-client:2.7.4")
+    api("com.zaxxer:HikariCP:5.0.0")
 }

--- a/common/src/main/java/org/kryonite/kryodiscordwhitelist/common/persistence/repository/impl/MariaDbUserRepository.java
+++ b/common/src/main/java/org/kryonite/kryodiscordwhitelist/common/persistence/repository/impl/MariaDbUserRepository.java
@@ -1,6 +1,6 @@
 package org.kryonite.kryodiscordwhitelist.common.persistence.repository.impl;
 
-import java.sql.Connection;
+import com.zaxxer.hikari.HikariDataSource;
 import java.sql.PreparedStatement;
 import java.sql.ResultSet;
 import java.sql.SQLException;
@@ -27,19 +27,19 @@ public class MariaDbUserRepository implements UserRepository {
       "UPDATE user SET minecraft_uuid = null, minecraft_name = ? "
           + "WHERE discord_id = ?";
 
-  private final Connection connection;
+  private final HikariDataSource dataSource;
 
-  public MariaDbUserRepository(Connection connection) throws SQLException {
-    this.connection = connection;
+  public MariaDbUserRepository(HikariDataSource dataSource) throws SQLException {
+    this.dataSource = dataSource;
 
-    try (PreparedStatement createTable = connection.prepareStatement(CREATE_USER_TABLE)) {
+    try (PreparedStatement createTable = dataSource.getConnection().prepareStatement(CREATE_USER_TABLE)) {
       createTable.executeUpdate();
     }
   }
 
   @Override
   public boolean addIfNotPresent(String minecraftName) throws SQLException {
-    try (PreparedStatement preparedStatement = connection.prepareStatement(GET_USER_BY_NAME)) {
+    try (PreparedStatement preparedStatement = dataSource.getConnection().prepareStatement(GET_USER_BY_NAME)) {
       preparedStatement.setString(1, minecraftName);
 
       ResultSet resultSet = preparedStatement.executeQuery();
@@ -54,7 +54,7 @@ public class MariaDbUserRepository implements UserRepository {
 
   @Override
   public void save(User user) throws SQLException {
-    try (PreparedStatement preparedStatement = connection.prepareStatement(GET_USER_BY_DISCORD_ID)) {
+    try (PreparedStatement preparedStatement = dataSource.getConnection().prepareStatement(GET_USER_BY_DISCORD_ID)) {
       preparedStatement.setLong(1, user.getDiscordId());
 
       ResultSet resultSet = preparedStatement.executeQuery();
@@ -67,7 +67,8 @@ public class MariaDbUserRepository implements UserRepository {
   }
 
   private void updateUser(User user) throws SQLException {
-    try (PreparedStatement preparedStatement = connection.prepareStatement(UPDATE_USER_WHERE_DISCORD_ID)) {
+    try (PreparedStatement preparedStatement = dataSource.getConnection()
+        .prepareStatement(UPDATE_USER_WHERE_DISCORD_ID)) {
       preparedStatement.setString(1, user.getMinecraftName());
       preparedStatement.setLong(2, user.getDiscordId());
 
@@ -76,7 +77,7 @@ public class MariaDbUserRepository implements UserRepository {
   }
 
   private void insertUser(long discordId, String minecraftName) throws SQLException {
-    try (PreparedStatement preparedStatement = connection.prepareStatement(INSERT_USER)) {
+    try (PreparedStatement preparedStatement = dataSource.getConnection().prepareStatement(INSERT_USER)) {
       preparedStatement.setLong(1, discordId);
       preparedStatement.setString(2, minecraftName);
 
@@ -86,7 +87,8 @@ public class MariaDbUserRepository implements UserRepository {
 
   @Override
   public boolean removeUser(String minecraftName) throws SQLException {
-    try (PreparedStatement preparedStatement = connection.prepareStatement(DELETE_USER_BY_MINECRAFT_NAME)) {
+    try (PreparedStatement preparedStatement = dataSource.getConnection()
+        .prepareStatement(DELETE_USER_BY_MINECRAFT_NAME)) {
       preparedStatement.setString(1, minecraftName);
 
       return preparedStatement.executeUpdate() > 0;
@@ -95,7 +97,8 @@ public class MariaDbUserRepository implements UserRepository {
 
   @Override
   public Optional<User> get(UUID minecraftUuid) throws SQLException {
-    try (PreparedStatement preparedStatement = connection.prepareStatement(GET_USER_BY_MINECRAFT_UUID)) {
+    try (
+        PreparedStatement preparedStatement = dataSource.getConnection().prepareStatement(GET_USER_BY_MINECRAFT_UUID)) {
       preparedStatement.setString(1, minecraftUuid.toString());
 
       ResultSet resultSet = preparedStatement.executeQuery();
@@ -115,7 +118,7 @@ public class MariaDbUserRepository implements UserRepository {
 
   @Override
   public boolean updateIfPresent(UUID minecraftUuid, String minecraftName) throws SQLException {
-    try (PreparedStatement preparedStatement = connection.prepareStatement(UPDATE_USER)) {
+    try (PreparedStatement preparedStatement = dataSource.getConnection().prepareStatement(UPDATE_USER)) {
       preparedStatement.setString(1, minecraftUuid.toString());
       preparedStatement.setString(2, minecraftName);
 

--- a/common/src/test/java/org/kryonite/kryodiscordwhitelist/common/persistence/repository/impl/MariaDbUserRepositoryTest.java
+++ b/common/src/test/java/org/kryonite/kryodiscordwhitelist/common/persistence/repository/impl/MariaDbUserRepositoryTest.java
@@ -17,7 +17,7 @@ import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
-import java.sql.Connection;
+import com.zaxxer.hikari.HikariDataSource;
 import java.sql.SQLException;
 import java.util.Optional;
 import java.util.UUID;
@@ -32,15 +32,15 @@ import org.mockito.junit.jupiter.MockitoExtension;
 class MariaDbUserRepositoryTest {
 
   @Mock(answer = Answers.RETURNS_DEEP_STUBS)
-  private Connection connection;
+  private HikariDataSource dataSource;
 
   @Test
   void shouldCreateTableOnStartup() throws SQLException {
     // Act
-    new MariaDbUserRepository(connection);
+    new MariaDbUserRepository(dataSource);
 
     // Assert
-    verify(connection).prepareStatement(CREATE_USER_TABLE);
+    verify(dataSource.getConnection()).prepareStatement(CREATE_USER_TABLE);
   }
 
   @Test
@@ -48,20 +48,20 @@ class MariaDbUserRepositoryTest {
     // Arrange
     String minecraftName = "Testee";
 
-    MariaDbUserRepository testee = new MariaDbUserRepository(connection);
-    when(connection.prepareStatement(GET_USER_BY_NAME).executeQuery().first()).thenReturn(false);
+    MariaDbUserRepository testee = new MariaDbUserRepository(dataSource);
+    when(dataSource.getConnection().prepareStatement(GET_USER_BY_NAME).executeQuery().first()).thenReturn(false);
 
     // Act
     boolean result = testee.addIfNotPresent(minecraftName);
 
     // Assert
     assertTrue(result, "Result should be true");
-    verify(connection, atLeastOnce()).prepareStatement(GET_USER_BY_NAME);
-    verify(connection.prepareStatement(GET_USER_BY_NAME)).setString(1, minecraftName);
+    verify(dataSource.getConnection(), atLeastOnce()).prepareStatement(GET_USER_BY_NAME);
+    verify(dataSource.getConnection().prepareStatement(GET_USER_BY_NAME)).setString(1, minecraftName);
 
-    verify(connection).prepareStatement(INSERT_USER);
-    verify(connection.prepareStatement(INSERT_USER)).setLong(1, -1);
-    verify(connection.prepareStatement(INSERT_USER)).setString(2, minecraftName);
+    verify(dataSource.getConnection()).prepareStatement(INSERT_USER);
+    verify(dataSource.getConnection().prepareStatement(INSERT_USER)).setLong(1, -1);
+    verify(dataSource.getConnection().prepareStatement(INSERT_USER)).setString(2, minecraftName);
   }
 
   @Test
@@ -69,81 +69,83 @@ class MariaDbUserRepositoryTest {
     // Arrange
     String minecraftName = "Testee";
 
-    MariaDbUserRepository testee = new MariaDbUserRepository(connection);
-    when(connection.prepareStatement(GET_USER_BY_NAME).executeQuery().first()).thenReturn(true);
+    MariaDbUserRepository testee = new MariaDbUserRepository(dataSource);
+    when(dataSource.getConnection().prepareStatement(GET_USER_BY_NAME).executeQuery().first()).thenReturn(true);
 
     // Act
     boolean result = testee.addIfNotPresent(minecraftName);
 
     // Assert
     assertFalse(result, "Result should be false");
-    verify(connection, atLeastOnce()).prepareStatement(GET_USER_BY_NAME);
-    verify(connection.prepareStatement(GET_USER_BY_NAME)).setString(1, minecraftName);
+    verify(dataSource.getConnection(), atLeastOnce()).prepareStatement(GET_USER_BY_NAME);
+    verify(dataSource.getConnection().prepareStatement(GET_USER_BY_NAME)).setString(1, minecraftName);
 
-    verify(connection, never()).prepareStatement(INSERT_USER);
+    verify(dataSource.getConnection(), never()).prepareStatement(INSERT_USER);
   }
 
   @Test
   void shouldSaveUser() throws SQLException {
     // Arrange
     User user = User.create(123456L, "Test", null);
-    MariaDbUserRepository testee = new MariaDbUserRepository(connection);
-    when(connection.prepareStatement(GET_USER_BY_DISCORD_ID).executeQuery().first()).thenReturn(false);
+    MariaDbUserRepository testee = new MariaDbUserRepository(dataSource);
+    when(dataSource.getConnection().prepareStatement(GET_USER_BY_DISCORD_ID).executeQuery().first()).thenReturn(false);
 
     // Act
     testee.save(user);
 
     // Assert
-    verify(connection, atLeastOnce()).prepareStatement(GET_USER_BY_DISCORD_ID);
-    verify(connection.prepareStatement(GET_USER_BY_DISCORD_ID)).setLong(1, user.getDiscordId());
+    verify(dataSource.getConnection(), atLeastOnce()).prepareStatement(GET_USER_BY_DISCORD_ID);
+    verify(dataSource.getConnection().prepareStatement(GET_USER_BY_DISCORD_ID)).setLong(1, user.getDiscordId());
 
-    verify(connection).prepareStatement(INSERT_USER);
-    verify(connection.prepareStatement(INSERT_USER)).setLong(1, user.getDiscordId());
-    verify(connection.prepareStatement(INSERT_USER)).setString(2, user.getMinecraftName());
+    verify(dataSource.getConnection()).prepareStatement(INSERT_USER);
+    verify(dataSource.getConnection().prepareStatement(INSERT_USER)).setLong(1, user.getDiscordId());
+    verify(dataSource.getConnection().prepareStatement(INSERT_USER)).setString(2, user.getMinecraftName());
   }
 
   @Test
   void shouldSaveAndUpdateUser_WhenAlreadyPresent() throws SQLException {
     // Arrange
     User user = User.create(123456L, "Test", null);
-    MariaDbUserRepository testee = new MariaDbUserRepository(connection);
-    when(connection.prepareStatement(GET_USER_BY_DISCORD_ID).executeQuery().first()).thenReturn(true);
+    MariaDbUserRepository testee = new MariaDbUserRepository(dataSource);
+    when(dataSource.getConnection().prepareStatement(GET_USER_BY_DISCORD_ID).executeQuery().first()).thenReturn(true);
 
     // Act
     testee.save(user);
 
     // Assert
-    verify(connection, atLeastOnce()).prepareStatement(GET_USER_BY_DISCORD_ID);
-    verify(connection.prepareStatement(GET_USER_BY_DISCORD_ID)).setLong(1, user.getDiscordId());
+    verify(dataSource.getConnection(), atLeastOnce()).prepareStatement(GET_USER_BY_DISCORD_ID);
+    verify(dataSource.getConnection().prepareStatement(GET_USER_BY_DISCORD_ID)).setLong(1, user.getDiscordId());
 
-    verify(connection).prepareStatement(UPDATE_USER_WHERE_DISCORD_ID);
-    verify(connection.prepareStatement(UPDATE_USER_WHERE_DISCORD_ID)).setString(1, user.getMinecraftName());
-    verify(connection.prepareStatement(UPDATE_USER_WHERE_DISCORD_ID)).setLong(2, user.getDiscordId());
+    verify(dataSource.getConnection()).prepareStatement(UPDATE_USER_WHERE_DISCORD_ID);
+    verify(dataSource.getConnection().prepareStatement(UPDATE_USER_WHERE_DISCORD_ID))
+        .setString(1, user.getMinecraftName());
+    verify(dataSource.getConnection().prepareStatement(UPDATE_USER_WHERE_DISCORD_ID)).setLong(2, user.getDiscordId());
   }
 
   @Test
   void shouldRemoveUser() throws SQLException {
     // Arrange
     String minecraftName = "Testee";
-    MariaDbUserRepository testee = new MariaDbUserRepository(connection);
-    when(connection.prepareStatement(DELETE_USER_BY_MINECRAFT_NAME).executeUpdate()).thenReturn(1);
+    MariaDbUserRepository testee = new MariaDbUserRepository(dataSource);
+    when(dataSource.getConnection().prepareStatement(DELETE_USER_BY_MINECRAFT_NAME).executeUpdate()).thenReturn(1);
 
     // Act
     boolean result = testee.removeUser(minecraftName);
 
     // Assert
     assertTrue(result, "Result should be true");
-    verify(connection, atLeastOnce()).prepareStatement(DELETE_USER_BY_MINECRAFT_NAME);
-    verify(connection.prepareStatement(DELETE_USER_BY_MINECRAFT_NAME)).setString(1, minecraftName);
+    verify(dataSource.getConnection(), atLeastOnce()).prepareStatement(DELETE_USER_BY_MINECRAFT_NAME);
+    verify(dataSource.getConnection().prepareStatement(DELETE_USER_BY_MINECRAFT_NAME)).setString(1, minecraftName);
   }
 
   @Test
   void shouldGetUser() throws SQLException {
     // Arrange
     UUID minecraftUuid = UUID.randomUUID();
-    MariaDbUserRepository testee = new MariaDbUserRepository(connection);
-    when(connection.prepareStatement(GET_USER_BY_MINECRAFT_UUID).executeQuery().first()).thenReturn(true);
-    when(connection.prepareStatement(GET_USER_BY_MINECRAFT_UUID).executeQuery().getString(anyString()))
+    MariaDbUserRepository testee = new MariaDbUserRepository(dataSource);
+    when(dataSource.getConnection().prepareStatement(GET_USER_BY_MINECRAFT_UUID).executeQuery().first())
+        .thenReturn(true);
+    when(dataSource.getConnection().prepareStatement(GET_USER_BY_MINECRAFT_UUID).executeQuery().getString(anyString()))
         .thenReturn(minecraftUuid.toString());
 
     // Act
@@ -152,24 +154,25 @@ class MariaDbUserRepositoryTest {
     // Assert
     assertTrue(result.isPresent(), "User was not present");
     assertEquals(minecraftUuid, result.get().getMinecraftUuid(), "Minecraft uuid did not match");
-    verify(connection, atLeastOnce()).prepareStatement(GET_USER_BY_MINECRAFT_UUID);
-    verify(connection.prepareStatement(anyString())).setString(1, minecraftUuid.toString());
+    verify(dataSource.getConnection(), atLeastOnce()).prepareStatement(GET_USER_BY_MINECRAFT_UUID);
+    verify(dataSource.getConnection().prepareStatement(anyString())).setString(1, minecraftUuid.toString());
   }
 
   @Test
   void shouldReturnEmptyOptional_WhenUserNotFound() throws SQLException {
     // Arrange
     UUID minecraftUuid = UUID.randomUUID();
-    MariaDbUserRepository testee = new MariaDbUserRepository(connection);
-    when(connection.prepareStatement(GET_USER_BY_MINECRAFT_UUID).executeQuery().first()).thenReturn(false);
+    MariaDbUserRepository testee = new MariaDbUserRepository(dataSource);
+    when(dataSource.getConnection().prepareStatement(GET_USER_BY_MINECRAFT_UUID).executeQuery().first())
+        .thenReturn(false);
 
     // Act
     Optional<User> result = testee.get(minecraftUuid);
 
     // Assert
     assertTrue(result.isEmpty(), "User was present but was not expected");
-    verify(connection, atLeastOnce()).prepareStatement(GET_USER_BY_MINECRAFT_UUID);
-    verify(connection.prepareStatement(anyString())).setString(1, minecraftUuid.toString());
+    verify(dataSource.getConnection(), atLeastOnce()).prepareStatement(GET_USER_BY_MINECRAFT_UUID);
+    verify(dataSource.getConnection().prepareStatement(anyString())).setString(1, minecraftUuid.toString());
   }
 
   @Test
@@ -177,17 +180,17 @@ class MariaDbUserRepositoryTest {
     // Arrange
     UUID minecraftUuid = UUID.randomUUID();
     String minecraftName = "Testee";
-    MariaDbUserRepository testee = new MariaDbUserRepository(connection);
-    when(connection.prepareStatement(UPDATE_USER).executeUpdate()).thenReturn(1);
+    MariaDbUserRepository testee = new MariaDbUserRepository(dataSource);
+    when(dataSource.getConnection().prepareStatement(UPDATE_USER).executeUpdate()).thenReturn(1);
 
     // Act
     boolean result = testee.updateIfPresent(minecraftUuid, minecraftName);
 
     // Assert
     assertTrue(result, "Result was not true");
-    verify(connection, atLeastOnce()).prepareStatement(UPDATE_USER);
-    verify(connection.prepareStatement(anyString())).setString(1, minecraftUuid.toString());
-    verify(connection.prepareStatement(anyString())).setString(2, minecraftName);
+    verify(dataSource.getConnection(), atLeastOnce()).prepareStatement(UPDATE_USER);
+    verify(dataSource.getConnection().prepareStatement(anyString())).setString(1, minecraftUuid.toString());
+    verify(dataSource.getConnection().prepareStatement(anyString())).setString(2, minecraftName);
   }
 
   @Test
@@ -195,16 +198,16 @@ class MariaDbUserRepositoryTest {
     // Arrange
     UUID minecraftUuid = UUID.randomUUID();
     String minecraftName = "Testee";
-    MariaDbUserRepository testee = new MariaDbUserRepository(connection);
-    when(connection.prepareStatement(UPDATE_USER).executeUpdate()).thenReturn(0);
+    MariaDbUserRepository testee = new MariaDbUserRepository(dataSource);
+    when(dataSource.getConnection().prepareStatement(UPDATE_USER).executeUpdate()).thenReturn(0);
 
     // Act
     boolean result = testee.updateIfPresent(minecraftUuid, minecraftName);
 
     // Assert
     assertFalse(result, "Result was true");
-    verify(connection, atLeastOnce()).prepareStatement(UPDATE_USER);
-    verify(connection.prepareStatement(anyString())).setString(1, minecraftUuid.toString());
-    verify(connection.prepareStatement(anyString())).setString(2, minecraftName);
+    verify(dataSource.getConnection(), atLeastOnce()).prepareStatement(UPDATE_USER);
+    verify(dataSource.getConnection().prepareStatement(anyString())).setString(1, minecraftUuid.toString());
+    verify(dataSource.getConnection().prepareStatement(anyString())).setString(2, minecraftName);
   }
 }

--- a/velocity/src/main/java/org/kryonite/kryodiscordwhitelist/velocity/KryoDiscordWhitelistPlugin.java
+++ b/velocity/src/main/java/org/kryonite/kryodiscordwhitelist/velocity/KryoDiscordWhitelistPlugin.java
@@ -8,8 +8,9 @@ import com.velocitypowered.api.event.Subscribe;
 import com.velocitypowered.api.event.proxy.ProxyInitializeEvent;
 import com.velocitypowered.api.plugin.Plugin;
 import com.velocitypowered.api.proxy.ProxyServer;
+import com.zaxxer.hikari.HikariConfig;
+import com.zaxxer.hikari.HikariDataSource;
 import java.io.IOException;
-import java.sql.Connection;
 import java.sql.DriverManager;
 import java.sql.SQLException;
 import java.util.List;
@@ -33,7 +34,7 @@ import org.mariadb.jdbc.Driver;
 public class KryoDiscordWhitelistPlugin {
 
   private final ProxyServer server;
-  private Connection connection;
+  private HikariDataSource hikariDataSource;
   private MessagingService messagingService;
 
   @Inject
@@ -53,8 +54,8 @@ public class KryoDiscordWhitelistPlugin {
 
     UserRepository userRepository;
     try {
-      setupConnection();
-      userRepository = new MariaDbUserRepository(connection);
+      setupHikariDataSource();
+      userRepository = new MariaDbUserRepository(hikariDataSource);
     } catch (SQLException exception) {
       log.error("Failed to setup UserRepository", exception);
       return;
@@ -81,10 +82,12 @@ public class KryoDiscordWhitelistPlugin {
     return messagingController;
   }
 
-  private void setupConnection() throws SQLException {
-    if (connection == null) {
+  private void setupHikariDataSource() throws SQLException {
+    if (hikariDataSource == null) {
       DriverManager.registerDriver(new Driver());
-      connection = DriverManager.getConnection(getEnv("CONNECTION_STRING"));
+      HikariConfig hikariConfig = new HikariConfig();
+      hikariConfig.setJdbcUrl(getEnv("CONNECTION_STRING"));
+      hikariDataSource = new HikariDataSource(hikariConfig);
     }
   }
 

--- a/velocity/src/test/java/org/kryonite/kryodiscordwhitelist/velocity/KryoDiscordWhitelistPluginTest.java
+++ b/velocity/src/test/java/org/kryonite/kryodiscordwhitelist/velocity/KryoDiscordWhitelistPluginTest.java
@@ -9,7 +9,7 @@ import static org.mockito.Mockito.when;
 
 import com.velocitypowered.api.event.proxy.ProxyInitializeEvent;
 import com.velocitypowered.api.proxy.ProxyServer;
-import java.sql.Connection;
+import com.zaxxer.hikari.HikariDataSource;
 import java.sql.SQLException;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
@@ -31,7 +31,7 @@ class KryoDiscordWhitelistPluginTest {
   private ProxyServer proxyServerMock;
 
   @Mock(answer = Answers.RETURNS_DEEP_STUBS)
-  private Connection connection;
+  private HikariDataSource hikariDataSource;
 
   @Mock
   private MessagingService messagingService;
@@ -53,7 +53,7 @@ class KryoDiscordWhitelistPluginTest {
   void shouldNotRegisterListener_WhenDatabaseNotAvailable() throws SQLException {
     // Arrange
     ProxyInitializeEvent proxyInitializeEvent = new ProxyInitializeEvent();
-    when(connection.prepareStatement(anyString())).thenThrow(SQLException.class);
+    when(hikariDataSource.getConnection().prepareStatement(anyString())).thenThrow(SQLException.class);
 
     // Act
     testee.onInitialize(proxyInitializeEvent);

